Changes to how data is stored

tbmiller-astro opened the pull request:
Changes to how data is stored within the `inference` class, discussed in #678. By default all the behavior should be the same as before but the changes help reduce the overall memory overhead and add extra flexibility when dealing with large datasets.

Summary of major changes:
- Simulated data is now stored in a torch `Dataset` rather than in lists as before. In `train()` only a call to `get_dataloaders` is needed.
- `get_dataloader` is changed to create the Dataloader from the saved Dataset. Takes the argument `start_round` to define which round(s) of simulations to load.
- All data checks have been moved to `append_simulations()`. Three arguments have been added: `return_self` to control if the method returns a copy of the class, `data_device` to control where data lives, independent of the device used for training and `warn_if_zscoring` to control if the z-scoring check is initiated.

janfb commented:
Hi @tbmiller-astro, 

thanks a lot for this PR, and sorry for the delay in the review--it seems we are all very busy at the moment. But I plan to have a look at this soon! 

Best, 
Jan

michaeldeistler commented:
Hi @tbmiller-astro,

thanks a lot for the PR. Your code is very clean and easy to follow, and your PR description also helped a lot! Thanks!

I like that you introduced a `data_device` and that all data checks are moved to `append_simulaltions()`.

I have two questions regarding your rational regarding other changes though:

1) Why is it preferable to store all data in a `torch.TensorDataset` instead of in `lists` of `tensors`? One could also have each `Tensor` in the `list` lie on the `data_device`, no? Is there a strong reason to directly concatenate the `Tensors` into a single `torch.TensorDataset` instead of doing this during `.train()`?

2) In what scenario would one use `append_simulations(..., return_self=False)`? You write that it `controls if the method returns a copy of the class`. However, since there is no `deepcopy(self)`, I am not sure if `return_self=True` actually doubles the memory footprint.

Thanks a lot for the PR!
Michael

michaeldeistler reviewed (changes requested):
Hi Tim,

thanks a lot for your answer, this makes sense! I left a few minor comments below.

I think that I would prefer to store the data as `lists` of `Tensors` instead of as `torch.TensorDataset`. Here's why:
1) It does not require [`get_simulations_indices()`](https://github.com/mackelab/sbi/pull/685/files#diff-d2b587a587456950f21d522076ff8cbf5b9aae1ae9dcc35e0e328ea7391d3302R352), but we can instead use the simpler [`get_simulations_since_round()`](https://github.com/mackelab/sbi/pull/685/files#diff-d2b587a587456950f21d522076ff8cbf5b9aae1ae9dcc35e0e328ea7391d3302L333)
2) Not all operations (e.g, z-scroing) require the `torch.TensorDataset`. So, I would prefer to generate the `torch.TensorDataset` only when it is really needed (i.e. in `.train()`).
3) My main reason is that `lists` provide a nice way of structuring data that was passed over several rounds. This structure is lost when all simulations are concatenated into a single `torch.TensorDataset`. Sure, one can potentially recover the structure via `_num_sims_per_round`, but it's just less easy to see what is happening.

So, to be clear, I would revert [these changes](https://github.com/mackelab/sbi/pull/685/files#diff-4835625eaf8d6e4ae66007ea13eb215b83cba4d5b96410b33301495163e2e5b1L180-L189) in `get_simulations()` and [these](https://github.com/mackelab/sbi/pull/685/files#diff-43883af88d1ab3b7811a1a061067c2a0aa1f70add166a298b59df50c6dfd34e6R139-R153) in `append_simulations()`. I am pefectly fine with your suggestion that the `get_dataloaders()` should, in a single call, generate the dataloader. However, I would suggest to generate the dataloaders from the lists instead of from the `torch.TensorDataset`.

Please let me know if you are up for implementing these changes! If you are busy, we can also merge your PR as it is right now and I'll make the changes myself.

Thanks again!

Best wishes
Michael

michaeldeistler commented on sbi/inference/base.py:
```diff
@@ -177,27 +176,24 @@ def get_simulations(
+
+        theta = torch.stack(theta)
+        x = torch.stack(x)
+        prior_masks = torch.stack(prior_masks).squeeze()
```
(but this part of the code will be removed anyways when reverting to `lists` of `Tensors`)

michaeldeistler commented on sbi/inference/base.py:
```diff
@@ -239,15 +235,20 @@ def get_dataloaders(
-        # Get total number of training examples.
-        num_examples = len(dataset)
+        # Generate indicies to use based on starting_round
+        indices = get_simulations_indcies(
```
(again, this part of the code will be removed anyways when reverting to lists of Tensors)

tbmiller-astro commented:
Hey @michaeldeistler, sounds good! Happy to make the changes suggested.

For the data checks, I've found they have a big affect on the memory footprint. Both the functions `warn_if_zscoring_changes_data` and `handle_invalid_x` create copies of the entire dataset. I could change it so that they are all under one option, like `run_data_checks` (or something similar) to reduce the number of arguments?

michaeldeistler commented:
Hi Tim,

awesome, thanks!

Hmmm I was not aware of this. However, `handle_invalid_data` is performed regardless of the `warn_on_invalid`, see [here](https://github.com/mackelab/sbi/pull/685/files#diff-43883af88d1ab3b7811a1a061067c2a0aa1f70add166a298b59df50c6dfd34e6R123). I do not think that there is a way around this, since invalid simulations must be detected and removed, otherwise training will fail.

So, since we already have to copy the dataset once (in `handle_invalid_data`), it should also be fine to copy it during `warn_if_zscoring_changes_data()`. Or am I missing something?

I agree that we should ensure that we have as few copies of the whole dataset as possible. I created an issue #695 to track this.

For now, I would suggest that we remove the option from `append_simulations()`. We might come back to this in a future PR, but I'd like to think about this a bit more.

tbmiller-astro commented:
Just addressed these requests. Also changed `.validate_x_and_theta()` to take `data_device` as an additional argument and produce more meaningful messages with these changes. - Tim

michaeldeistler commented on sbi/inference/snle/snle_base.py:
```diff
@@ -176,10 +188,14 @@ def train(
         if self._neural_net is None or retrain_from_scratch:
+
+            # Get theta,x from dataset to initialize NN
+            theta, x, _ = self.get_simulations()
```
I think this should be 
```
theta, x, _ = self.get_simulations(starting_round=start_idx)
```

michaeldeistler commented on sbi/inference/snpe/snpe_base.py:
```diff
@@ -235,45 +262,39 @@ def train(
+            x = self._x_roundwise[0][:training_batch_size]
+            theta = self._theta_roundwise[0][:training_batch_size]
+            self._neural_net = self._build_neural_net(theta.to("cpu"), x.to("cpu"))
+            self._x_shape = x_shape_from_simulation(x.to("cpu"))
```
Could you replace these lines with the code that you also used in `snle_base.py` and `snre_base.py` (i.e. use `get_simulations`)? Or is there a reason to not use `get_simulations()` here?

michaeldeistler commented on sbi/inference/snre/snre_base.py:
```diff
@@ -183,11 +198,14 @@ def train(
         if self._neural_net is None or retrain_from_scratch:
+
+            # Get theta,x from dataset to initialize NN
+            theta, x, _ = self.get_simulations()
```
I think this should be
```
theta, x, _ = self.get_simulations(starting_round=start_idx)
```

michaeldeistler reviewed (approved):
This looks great! Thanks so much for your taking the time to fix these things and going through the review process! I left very minor comments (but feel free to ignore them if you are busy). I'll merge the PR on Monday.

All the best and thanks again!
Michael

janfb commented on sbi/utils/user_input_checks.py:
```diff
@@ -647,7 +647,7 @@ def check_estimator_arg(estimator: Union[str, Callable]) -> None:
+    theta: Any, x: Any, data_device: str = "cpu", training_device: str = "cpu"
 ) -> Tuple[Tensor, Tensor]:
     r"""
     Checks if the passed $(\theta, x)$ are valid.
```
I think it would be great to update this docstring and be more explicit about what this function is checking. What's the difference between training and data device, whats the overall goal of this function?

janfb reviewed (commented):
Overall great! Thanks a lot for working on this! 
see comments below.

tbmiller-astro commented:
Thanks both! I should be able to make these final changes tomorrow! - Tim

michaeldeistler commented:
Thanks again! This is really fantastic work!

All the best
Michael
